### DISCUSSION TITLE
10.1: Track time from records with interactions

### DISCUSSION THREAD

**Discussion by jstanden** [cerb/cerb-release#38] [Guides and Tutorials]
# Introduction

In the Cerb [10.0](https://cerb.ai/releases/10.0/) update the built-in time tracking option was removed from ticket profiles.

In this guide you'll learn how to add this functionality back as an [interaction](https://cerb.ai/docs/interactions/).

# Import the package in Cerb

Log into Cerb as an administrator.

Navigate to **Setup -> Packages -> Import**.

Paste the following package:

```json
{
  "package": {
    "name": "Time tracking interaction example"
  },
  "records": [
    {
      "uid": "automation_track_time",
      "_context": "automation",
      "name": "wgm.example.timetracking",
      "description": "Create time tracking entries from record profiles",
      "extension_id": "cerb.trigger.interaction.worker",
      "script": "inputs:\n  text/record_type:\n    required@bool: yes\n    type: record_type\n  text/record_id:\n    required@bool: yes\n    type: number\n\nstart:\n  set:\n    record__context: {{inputs.record_type}}\n    record_id: {{inputs.record_id}}\n    \n  await/record:\n    record:\n      uri: cerb:time_tracking\n      output: new_timeentry\n  \n  # See: https://cerb.ai/docs/records/types/time_entry/\n  record.update:\n    output: new_timeentry\n    inputs:\n      record_type: timetracking\n      record_id: {{new_timeentry.record.id}}\n      fields:\n        links@trim,list:\n          {{record__context}}:{{record_id}}\n          {% if \n            record__context is record type ('ticket') \n            and record_org_id\n          %}\n          org:{{record_org_id}}\n          {% endif %}\n",
      "policy_kata": "commands:\n  record.update:\n    deny/type@bool: {{inputs.record_type is not record type ('timetracking')}}\n    deny/fields@bool: {{inputs.fields|keys != ['links']}}\n    allow@bool: yes"
    }
  ],
  "toolbars": [
    {
      "toolbar": "record.profile",
      "kata": "# Show time tracking on ticket profiles\ninteraction/trackTime:\n  uri: cerb:automation:wgm.example.timetracking\n  tooltip: Track Time\n  icon: stopwatch\n  hidden@bool:\n    {{\n      record__type is not record type ('ticket')\n    }}\n  inputs:\n    record_type@key: record__context\n    record_id@key: record_id\n  after:\n    refresh_widgets@csv: Ticket"
    }
  ]
}
```

Click the **Import** button.

You should see the following output.

<kbd><img width="286" alt="image" src="https://user-images.githubusercontent.com/63288/126725057-661c2a6d-3c72-4058-a08e-ba8a6c2bb7f1.png"></kbd>

# Tracking time

Navigate to any ticket profile from **Search -> Tickets**.

You'll see a time tracking icon in the profile toolbar:

<kbd>![image](https://user-images.githubusercontent.com/63288/126725161-f80a2803-2449-45e4-912f-7840fc018408.png)</kbd>

Clicking this will open an interaction to create a time tracking entry:

<kbd><img width="670" alt="image" src="https://user-images.githubusercontent.com/63288/126725192-dc10ac9c-599c-4fdb-8b0e-4c3b1e14f146.png"></kbd>

You can either manually enter a duration to record it, or click the **Resume** button to continue the timer:

<kbd><img width="313" alt="image" src="https://user-images.githubusercontent.com/63288/126725247-6f0d175d-107a-463b-8ae8-5efaedffd560.png"></kbd>

# Conclusion

Workers can now quickly create time tracking entries from ticket profiles. You can modify the [record.profile](https://cerb.ai/docs/toolbars/interactions/record.profile/) [toolbar](https://cerb.ai/docs/toolbars/) to allow time tracking from other [record types](https://cerb.ai/docs/records/types/).

**Comment by rp-sg** [comment]
Thanks for this as this has helped restore the ticket profile time tracking functionality on our recent jump to cerb 10.3.7!

I am looking into whether we can better replicate the previous behavior that our staff is used to in which the time tracking toolbar button starts the timer, and the stop timer button is what opens the time tracking popup to finalize the tracking entry. The log date on the entry would be set to whenever the timer was stopped. If this isn't possible, whenever the 'resume' button timer is used would it be possible to have the pre-filled log date update to the time in which the timer was stopped? Any assistance or pointers to achieve this would be appreciated. Thanks in advance!

**Reply by jstanden** [reply]
I'll implement that in 10.3.8 for you.

In the interaction toolbar, we could just have a syntax like:

```yaml
start:
  record.create/entry:
    output: new_timeentry
    inputs:
      record_type: time_entry
      fields:
        secs: 0
        worker_id@int: {{worker_id}}
  return:
    timer: {{new_timeentry.id}}
```

That would create a new time tracking entry to log into, then start the timer. It wouldn't require hitting 'Resume' in the editor popup like the example in this guide.

**Comment by rp-sg** [comment]
Thanks Jeff, much appreciated.
